README.md
# LazyMapper

Wraps a Hash and lazily maps its attributes to rich domain objects using either a set of default mappers (for Ruby's built-in types), or custom mappers specified by the client.

The mapped values are memoized.

Example:

    class Foo < LazyMapper::Model
      one :id, Integer, from: 'iden'
      one :created_at, Time
      one :amount, Money, map: Money.method(:parse)
      many :users, User, map: ->(u) { User.new(u) }
    end

## Documentation

See [RubyDoc](https://www.rubydoc.info/gems/lazy_mapper/0.4.0)

## License

See LICENSE file.

## Changes

See CHANGES.md


CHANGELOG.md
#0.4.0

## Changed

* Removed deprecated factory method `.from_json`
* Enable inheritance of mappers added after the derived class is defined

#0.3.2

## Fixed

* No more BigDecimal deprecation warning

#0.3.1

## Fixed

* Collections no longer missing from `#to_h` and `#inspect`

#0.3.0

## Changed
* `.from_json` renamed to `.from`. `.from_json` is kept as an alias to the new method.

## Added

* `#to_h` method

#0.2.1

(No user visible changes)



spec/spec_helper.rb
# frozen_string_literal: true

if ENV['RCOV']
  require 'simplecov'
  SimpleCov.start do
    add_filter '/spec/'
  end
end

RSpec.configure do |config|
  config.pattern = '**{,/*/**}/*{_,.}spec.rb'
end


spec/lazy_mapper_spec.rb
# frozen_string_literal: true

require 'spec_helper'
require 'lazy_mapper'

describe LazyMapper do

  describe 'when constructed from unmapped data' do

    subject(:instance) { klass.from unmapped_data }

    let(:unmapped_data) {
      {
        'createdAt' => '2015-07-27',
        'updatedAt' => ['2015-01-01', '2015-01-02'],
        'foo'       => '42',
        'blue'      => true
      }
    }
    let(:klass) {
      t = type
      m = map
      Class.new described_class do
        one :created_at, Date
        many :updated_at, Date
        one :foo, t, map: m, default: 666
        is :blue?
      end
    }
    let(:mapper) { spy 'mapper', map: 42 }
    let(:map) { ->(x) { mapper.map(x) } }
    let(:type) { Integer }

    context 'if the supplied data is nil' do
      let(:unmapped_data) { nil }

      it { is_expected.to be_nil }
    end

    context 'when invalid data is supplied' do
      let(:unmapped_data) { 'not a hash' }

      it 'fails with a TypeError' do
        expect { instance }.to raise_error(TypeError)
      end
    end

    context 'when valid data is supplied' do

      it 'maps primitives to domain objects' do
        expect(instance.created_at).to eq(Date.new(2015, 7, 27))
      end

      it 'maps arrays of primitives to arrays of domain objects' do
        expect(instance.updated_at).to be_a(Array)
        expect(instance.updated_at.first).to be_a(Date)
        expect(instance).to be_blue
      end

      it 'memoizes mapped value so that potentially expensive mappings are performed just once' do
        3.times do
          expect(instance.foo).to eq(42)
        end
        expect(mapper).to have_received(:map).exactly(1).times.with('42')
      end

      it 'only shows already mapped values, when inspected' do
        stub_const 'MyModel', klass
        instance.created_at
        expect(instance.inspect).to eq '<MyModel created_at: #<Date: 2015-07-27 ((2457231j,0s,0n),+0s,2299161j)> >'
      end

      context 'if the mapped value is nil' do
        let(:map) { -> x { mapper.map(x); nil } }

        it 'even memoizes that' do
          3.times do
            expect(instance.foo).to be_nil
          end
          expect(mapper).to have_received(:map).exactly(1).times.with('42')
        end
      end

      context 'when the model has circular references' do

        subject(:instance) { foo }

        let(:foo) { klass_foo.from 'bar' => 'bar' }
        let(:bar) { klass_bar.from 'foo' => 'foo' }
        let(:foo_builder) { proc { foo } }
        let(:bar_builder) { proc { bar } }

        let(:klass_foo) {
          b = bar_builder
          Class.new described_class do
            one :bar, Object, map: b
          end
        }

        let(:klass_bar) {
          b = foo_builder
          Class.new described_class do
            one :foo, Object, map: b
          end
        }

        it 'avoids infinit recursion, when inspected' do
          stub_const('Bar', klass_bar)
          stub_const('Foo', klass_foo)
          foo.bar.foo
          expect(foo.inspect).to eq('<Foo bar: <Bar foo: <Foo ... > > >')
        end
      end
    end

    describe 'the :from option' do

      let(:klass) {
        Class.new described_class do
          one :baz, Integer, from: 'BAZ'
          is :fuzzy?, from: 'hairy'
          is :sweet?, from: 'sugary'
        end
      }

      let(:unmapped_data) { { 'BAZ' => 999, 'hairy' => true } }

      it 'specifies the name of the attribute in the unmapped data' do
        expect(instance.baz).to eq(999)
      end

      it { is_expected.to be_fuzzy }
      it { is_expected.to_not be_sweet }
    end

    context "if the mapper doesn't map to the correct type" do

      let(:klass) {
        Class.new described_class do
          one :bar, Float, map: ->(x) { x.to_s }
        end
      }

      it 'fails with a TypeError when an attribute is accessed' do
        instance = klass.from 'bar' => 42
        expect { instance.bar }.to raise_error(TypeError)
      end
    end

    it 'supports adding custom type mappers to instances' do
      type = Struct.new(:val1, :val2)
      klass = Class.new described_class do
        one :composite, type
      end

      instance = klass.from 'composite' => '123 456'
      instance.add_mapper_for(type) { |unmapped_value| type.new(*unmapped_value.split(' ')) }

      expect(instance.composite).to eq type.new('123', '456')

      instance = klass.new composite: type.new('abc', 'cde')
      expect(instance.composite).to eq type.new('abc', 'cde')
    end

    it 'supports injection of customer mappers during instantiation' do
      type = Struct.new(:val1, :val2)
      klass = Class.new described_class do
        one :foo, type
        one :bar, type
      end

      instance = klass.from({ 'foo' => '123 456', 'bar' => 'abc def' },
                            mappers: {
                              foo: ->(f) { type.new(*f.split(' ').reverse) },
                              type => ->(t) { type.new(*t.split(' ')) }
                            })

      expect(instance.foo).to eq type.new('456', '123')
      expect(instance.bar).to eq type.new('abc', 'def')
    end

    it 'expects the supplied mapper to return an Array if the unmapped value of a "many" attribute is not an array' do
      klass = Class.new described_class do
        many :foos, String, map: ->(v) { return v.split '' }
        many :bars, String, map: ->(v) { return v }
      end

      instance = klass.from 'foos' => 'abc', 'bars' => 'abc'

      expect(instance.foos).to eq %w[a b c]
      expect { instance.bars }.to raise_error(TypeError)
    end

    context 'when it is derived from another LazyMapper' do
      let(:klass) { Class.new(base) }
      let(:composite_type) { Struct.new(:val1, :val2) }
      let(:new_type) { Class.new }
      let(:new_type_mapper) { new_type.method(:new) }
      let(:base) {
        type = composite_type
        Class.new(described_class) do
          default_value_for type, type.new('321', '123')
          mapper_for type, ->(unmapped_value) { type.new(*unmapped_value.split(' ')) }
          one :composite, type
        end
      }

      it 'inherits attributes' do
        expect(klass.attributes.keys).to eq [:composite]
        expect(instance).to respond_to(:composite)
      end

      it 'inherits default values' do
        expect(instance.composite).to eq composite_type.new('321', '123')
      end

      it 'inherits default mappers' do
        expect(klass.from('composite' => 'abc def').composite).to eq composite_type.new('abc', 'def')
      end

      it 'inherits default mappers that are added to its parent after it has been defined', wip: true do
        expect(klass.mappers[new_type]).to be_nil
        base.mapper_for new_type, new_type_mapper
        expect(base.mappers[new_type]).to eq new_type_mapper
        expect(klass.mappers[new_type]).to eq new_type_mapper
      end
    end
  end

  context 'when constructed with .new' do

    subject(:instance) { klass.new values }
    let(:values) { {} }

    let(:klass) {
      Class.new described_class do
        one :title, String
        one :count, Integer
        one :rate, Float
        one :tags, Array
        one :widget, Object
        one :things, Array, default: ['something']
        is  :green?
        has :flowers?
        many :cars, Object
      end
    }

    context 'when values are provided' do

      let(:values) {
        {
          title:  'A title',
          count:  42,
          rate:   3.14,
          tags:   %w[red hot],
          widget: Date.new,
          things: %i[one two three],
          green?: true
        }
      }

      it 'uses those values' do
        values.each do |name, value|
          expect(instance.send(name)).to eq(value)
        end
      end

      context 'if a value given in the constructor is not of the specified type' do
        let(:values) {
          { title: :'Not a string' }
        }

        it 'fails with a TypeError' do
          expect { instance }.to raise_error(TypeError)
        end
      end
    end

    context 'when no values are provided' do

      it 'has sensible default values for primitive types' do
        expect(instance.title).to eq('')
        expect(instance.count).to eq(0)
        expect(instance.rate).to eq(0.0)
        expect(instance.widget).to be_nil
        expect(instance.tags).to eq []
      end

      it 'uses the supplied default values' do
        expect(instance.things).to eq(['something'])
      end

      it 'falls back to nil in all other cases' do
        expect(instance.widget).to be_nil
      end

      it 'dosn\'t share its default values with other instances' do
        instance1 = klass.new
        instance2 = klass.new
        instance1.tags << 'dirty'
        instance1.things.pop
        expect(instance2.tags).to be_empty
        expect(instance2.things).to_not be_empty
      end

      it 'still includes every attribute when converted to Hash' do
        expect(instance.to_h).to eq(
          title:    '',
          count:    0,
          rate:     0.0,
          tags:     [],
          widget:   nil,
          things:   ['something'],
          green?:   false,
          flowers?: false,
          cars:     []
        )
      end
    end
  end
end


lib/lazy_mapper/lazy_mapper.rb
# frozen_string_literal: true

#
# Wraps a Hash or Hash-like data structure of primitive values and lazily maps
# its attributes to semantically rich domain objects using either a set of
# default mappers (for Ruby's built-in value types), or custom mappers which
# can be added either at the class level or at the instance level.
#
# Example:
#     class Foo < LazyMapper
#       one :id, Integer, from: 'xmlId'
#       one :created_at, Time
#       one :amount, Money, map: Money.method(:parse)
#       many :users, User, map: ->(u) { User.new(u) }
#     end
#
class LazyMapper

  #
  # Adds (or overrides) a default type for a given type
  #
  def self.default_value_for type, value
    default_values[type] = value
  end

  def self.default_values
    @default_values ||= DEFAULT_VALUES
  end

  #
  # Adds a mapper for a give type
  #
  def self.mapper_for(type, mapper)
    mappers[type] = mapper
  end

  def self.mappers
    @mappers ||= DEFAULT_MAPPINGS
  end

  def self.attributes
    @attributes ||= {}
  end

  def self.inherited klass # :nodoc:
    # Make the subclass "inherit" the values of these class instance variables
    %i[
      default_values
      attributes
    ].each do |s|
      klass.instance_variable_set IVAR[s], self.send(s).dup
    end

    # If a mapper is does not exist in the derived class, look it up in this class
    klass.instance_variable_set('@mappers', Hash.new { |_mappers, type| mappers[type] })
  end

  def mappers
    @mappers ||= self.class.mappers
  end

  IVAR = lambda { |name| # :nodoc:
    name_as_str = name.to_s
    name_as_str = name_as_str[0...-1] if name_as_str[-1] == '?'

    ('@' + name_as_str).freeze
  }

  WRITER = -> name { (name.to_s.delete('?') + '=').to_sym }

  #
  # Creates a new instance by giving a Hash of attribues.
  #
  # Attribute values are type checked according to how they were defined.
  #
  # Fails with +TypeError+, if a value doesn't have the expected type.
  #
  # == Example
  #
  #     Foo.new :id => 42,
  #       :created_at => Time.parse("2015-07-29 14:07:35 +0200"),
  #       :amount => Money.parse("$2.00"),
  #       :users => [
  #         User.new("id" => 23, "name" => "Adam"),
  #         User.new("id" => 45, "name" => "Ole"),
  #         User.new("id" => 66, "name" => "Anders"),
  #         User.new("id" => 91, "name" => "Kristoffer)
  #       ]

  def initialize(values = {})
    @mappers = {}
    values.each do |name, value|
      send(WRITER[name], value)
    end
  end

  #
  # Create a new instance by giving a Hash of unmapped attributes.
  #
  # The keys in the Hash are assumed to be camelCased strings.
  #
  # == Arguments
  #
  # +unmapped_data+ - The unmapped data as a Hash(-like object). Must respond to #to_h.
  # Keys are assumed to be camelCased string
  #
  # +mappers:+ - Optional instance-level mappers.
  # Keys can either be classes or symbols corresponding to named attributes.
  #
  #
  # == Example
  #
  #     Foo.from({
  #       "xmlId" => 42,
  #       "createdAt" => "2015-07-29 14:07:35 +0200",
  #       "amount" => "$2.00",
  #       "users" => [
  #         { "id" => 23, "name" => "Adam" },
  #         { "id" => 45, "name" => "Ole" },
  #         { "id" => 66, "name" => "Anders" },
  #         { "id" => 91, "name" => "Kristoffer" } ]},
  #       mappers: {
  #         :amount => -> x { Money.new(x) },
  #         User    => User.method(:new) })
  #
  def self.from unmapped_data, mappers: {}
    return nil if unmapped_data.nil?
    fail TypeError, "#{ unmapped_data.inspect } is not a Hash" unless unmapped_data.respond_to? :to_h

    instance = new
    instance.send :unmapped_data=, unmapped_data.to_h
    instance.send :mappers=, mappers
    instance
  end

  #
  # Defines an attribute and creates a reader and a writer for it.
  # The writer verifies the type of it's supplied value.
  #
  # == Arguments
  #
  # +name+ - The name of the attribue
  #
  # +type+ - The type of the attribute. If the wrapped value is already of that type, the mapper is bypassed.
  # If the type is allowed be one of several, use an Array to to specify which ones
  #
  # +from:+ - Specifies the name of the wrapped value in the JSON object. Defaults to camelCased version of +name+.
  #
  # +map:+ - Specifies a custom mapper to apply to the wrapped value.
  # If unspecified, it defaults to the default mapper for the specified +type+ or simply the identity mapper
  # if no default mapper exists.
  #
  # +default:+ - The default value to use, if the wrapped value is not present in the wrapped JSON object.
  #
  # +allow_nil:+ - If true, allows the mapped value to be nil. Defaults to true.
  #
  # == Example
  #
  #    class Foo < LazyMapper
  #      one :boss, Person, from: "supervisor", map: ->(p) { Person.new(p) }
  #      one :weapon, [BladedWeapon, Firearm], default: Sixshooter.new
  #      # ...
  #    end
  #
  def self.one(name, type, from: map_name(name), allow_nil: true, **args)

    ivar = IVAR[name]

    # Define writer
    define_method(WRITER[name]) { |val|
      check_type! val, type, allow_nil: allow_nil
      instance_variable_set(ivar, val)
    }

    # Define reader
    define_method(name) {
      memoize(name, ivar) {
        unmapped_value = unmapped_data[from]
        mapped_value(name, unmapped_value, type, **args)
      }
    }

    attributes[name] = type
  end

  #
  # Converts a value to +true+ or +false+ according to its truthyness
  #
  TO_BOOL = -> b { !!b }

  #
  # Defines an boolean attribute
  #
  # == Arguments
  #
  # +name+ - The name of the attribue
  #
  # +from:+ - Specifies the name of the wrapped value in the JSON object.
  # Defaults to camelCased version of +name+.
  #
  # +map:+ - Specifies a custom mapper to apply to the wrapped value. Must be a Callable.
  # Defaults to TO_BOOL if unspecified.
  #
  # +default:+ The default value to use if the value is missing. False, if unspecified
  #
  # == Example
  #
  #    class Foo < LazyMapper
  #      is :green?, from: "isGreen", map: ->(x) { !x.zero? }
  #      # ...
  #    end
  #
  def self.is name, from: map_name(name), map: TO_BOOL, default: false
    one name, [TrueClass, FalseClass], from: from, allow_nil: false, map: map, default: default
  end

  singleton_class.send(:alias_method, :has, :is)

  #
  # Defines a collection attribute
  #
  # == Arguments
  #
  # +name+ - The name of the attribute
  #
  # +type+ - The type of the elements in the collection.
  #
  # +from:+ - Specifies the name of the wrapped array in the unmapped data.
  # Defaults to camelCased version of +name+.
  #
  # +map:+ - Specifies a custom mapper to apply to each elements in the wrapped collection.
  # If unspecified, it defaults to the default mapper for the specified +type+ or simply the identity mapper
  # if no default mapper exists.
  #
  # +default:+ - The default value to use, if the unmapped value is missing.
  #
  # == Example
  #
  #    class Bar < LazyMapper
  #      many :underlings, Person, from: "serfs", map: ->(p) { Person.new(p) }
  #      # ...
  #    end
  #
  def self.many(name, type, from: map_name(name), **args)

    # Define setter
    define_method(WRITER[name]) { |val|
      check_type! val, Enumerable, allow_nil: false
      instance_variable_set(IVAR[name], val)
    }

    # Define getter
    define_method(name) {
      memoize(name) {
        unmapped_value = unmapped_data[from]
        if unmapped_value.is_a? Array
          unmapped_value.map { |v| mapped_value(name, v, type, **args) }
        else
          mapped_value name, unmapped_value, Array, **args
        end
      }
    }

    attributes[name] = Array
  end

  #
  # Adds an instance-level type mapper
  #
  def add_mapper_for(type, &block)
    mappers[type] = block
  end

  #
  # Returns a +Hash+ with keys corresponding to attribute names, and values
  # corresponding to *mapped* attribute values.
  #
  # Note: This will eagerly map all attributes that haven't yet been mapped
  #
  def to_h
    attributes.each_with_object({}) { |(key, _value), h|
      h[key] = self.send key
    }
  end

  def inspect
    @__under_inspection__ ||= 0
    return "<#{ self.class.name } ... >" if @__under_inspection__.positive?

    @__under_inspection__ += 1
    present_attributes = attributes.keys.each_with_object({}) { |name, memo|
      ivar = IVAR[name]
      next unless self.instance_variable_defined? ivar

      memo[name] = self.instance_variable_get ivar
    }

    res = "<#{ self.class.name } #{ present_attributes.map { |k, v| k.to_s + ': ' + v.inspect }.join(', ') } >"
    @__under_inspection__ -= 1
    res
  end

  #
  # Defines how to map an attribute name
  # to the corresponding name in the unmapped
  # JSON object.
  #
  # Defaults to CAMELIZE
  #
  def self.map_name(name)
    CAMELIZE[name]
  end

  private

  attr_writer :unmapped_data
  attr_writer :mappers

  def unmapped_data
    @unmapped_data ||= {}
  end

  def mapping_for(name, type)
    mappers[name] || mappers[type] || self.class.mappers[type]
  end

  def default_value(type)
    self.class.default_values[type]
  end

  def attributes
    self.class.attributes
  end

  def mapped_value(name, unmapped_value, type, map: mapping_for(name, type), default: default_value(type))
    return default.dup if unmapped_value.nil? # Duplicate to prevent accidental sharing between instances

    if map.nil?
      fail ArgumentError, "missing mapper for #{ name } (#{ type }). "\
        "Unmapped value: #{ unmapped_value.inspect }"
    end

    return map.call(unmapped_value, self) if map.arity > 1

    map.call(unmapped_value)
  end

  def check_type! value, type, allow_nil:
    permitted_types = allow_nil ? Array(type) + [ NilClass ] : Array(type)
    return if permitted_types.any? value.method(:is_a?)

    fail TypeError.new "#{ self.class.name }: "\
      "#{ value.inspect } is a #{ value.class } "\
      "but was supposed to be a #{ humanize_list permitted_types, conjunction: ' or ' }"
  end

  # [1,2,3] -> "1, 2 and 3"
  # [1, 2]  -> "1 and 2"
  # [1]     -> "1"
  def humanize_list terms, separator: ', ', conjunction: ' and '
    *all_but_last, last = terms
    return last if all_but_last.empty?

    [ all_but_last.join(separator), last ].join conjunction
  end

  def memoize name, ivar = IVAR[name]
    send WRITER[name], yield unless instance_variable_defined?(ivar)
    instance_variable_get(ivar)
  end

  SNAKE_CASE_PATTERN = /(_[a-z])/ # :nodoc:
  CAMELIZE = -> name { name.to_s.gsub(SNAKE_CASE_PATTERN) { |x| x[1].upcase }.delete('?') }

end


lib/lazy_mapper/model.rb
# frozen_string_literal: true

module LazyMapper

  #
  # Wraps a Hash or Hash-like data structure of primitive values and lazily maps
  # its attributes to semantically rich domain objects using either a set of
  # default mappers (for Ruby's built-in value types), or custom mappers which
  # can be added either at the class level or at the instance level.
  #
  # Example:
  #     class Foo < LazyMapper::Model
  #       one :id, Integer, from: 'xmlId'
  #       one :created_at, Time
  #       one :amount, Money, map: Money.method(:parse)
  #       many :users, User, map: ->(u) { User.new(u) }
  #     end
  #
  class Model

    #
    # Adds (or overrides) a default type for a given type
    #
    def self.default_value_for type, value
      default_values[type] = value
    end

    def self.default_values
      @default_values ||= DEFAULT_VALUES
    end

    #
    # Adds a mapper for a give type
    #
    def self.mapper_for(type, mapper)
      mappers[type] = mapper
    end

    def self.mappers
      @mappers ||= DEFAULT_MAPPINGS
    end

    def self.attributes
      @attributes ||= {}
    end

    def self.inherited klass # :nodoc:
      # Make the subclass "inherit" the values of these class instance variables
      %i[
        default_values
        attributes
      ].each do |s|
        klass.instance_variable_set IVAR[s], self.send(s).dup
      end

      # If a mapper is does not exist in the derived class, look it up in this class
      klass.instance_variable_set('@mappers', Hash.new { |_mappers, type| mappers[type] })
    end

    def mappers
      @mappers ||= self.class.mappers
    end

    IVAR = lambda { |name| # :nodoc:
      name_as_str = name.to_s
      name_as_str = name_as_str[0...-1] if name_as_str[-1] == '?'

      ('@' + name_as_str).freeze
    }

    WRITER = -> name { (name.to_s.delete('?') + '=').to_sym }

    #
    # Creates a new instance by giving a Hash of attribues.
    #
    # Attribute values are type checked according to how they were defined.
    #
    # Fails with +TypeError+, if a value doesn't have the expected type.
    #
    # == Example
    #
    #     Foo.new :id => 42,
    #       :created_at => Time.parse("2015-07-29 14:07:35 +0200"),
    #       :amount => Money.parse("$2.00"),
    #       :users => [
    #         User.new("id" => 23, "name" => "Adam"),
    #         User.new("id" => 45, "name" => "Ole"),
    #         User.new("id" => 66, "name" => "Anders"),
    #         User.new("id" => 91, "name" => "Kristoffer)
    #       ]

    def initialize(values = {})
      @mappers = {}
      values.each do |name, value|
        send(WRITER[name], value)
      end
    end

    #
    # Create a new instance by giving a Hash of unmapped attributes.
    #
    # The keys in the Hash are assumed to be camelCased strings.
    #
    # == Arguments
    #
    # +unmapped_data+ - The unmapped data as a Hash(-like object). Must respond to #to_h.
    # Keys are assumed to be camelCased string
    #
    # +mappers:+ - Optional instance-level mappers.
    # Keys can either be classes or symbols corresponding to named attributes.
    #
    #
    # == Example
    #
    #     Foo.from({
    #       "xmlId" => 42,
    #       "createdAt" => "2015-07-29 14:07:35 +0200",
    #       "amount" => "$2.00",
    #       "users" => [
    #         { "id" => 23, "name" => "Adam" },
    #         { "id" => 45, "name" => "Ole" },
    #         { "id" => 66, "name" => "Anders" },
    #         { "id" => 91, "name" => "Kristoffer" } ]},
    #       mappers: {
    #         :amount => -> x { Money.new(x) },
    #         User    => User.method(:new) })
    #
    def self.from unmapped_data, mappers: {}
      return nil if unmapped_data.nil?
      fail TypeError, "#{ unmapped_data.inspect } is not a Hash" unless unmapped_data.respond_to? :to_h

      instance = new
      instance.send :unmapped_data=, unmapped_data.to_h
      instance.send :mappers=, mappers
      instance
    end

    #
    # Defines an attribute and creates a reader and a writer for it.
    # The writer verifies the type of it's supplied value.
    #
    # == Arguments
    #
    # +name+ - The name of the attribue
    #
    # +type+ - The type of the attribute. If the wrapped value is already of that type, the mapper is bypassed.
    # If the type is allowed be one of several, use an Array to to specify which ones
    #
    # +from:+ - Specifies the name of the wrapped value in the JSON object. Defaults to camelCased version of +name+.
    #
    # +map:+ - Specifies a custom mapper to apply to the wrapped value.
    # If unspecified, it defaults to the default mapper for the specified +type+ or simply the identity mapper
    # if no default mapper exists.
    #
    # +default:+ - The default value to use, if the wrapped value is not present in the wrapped JSON object.
    #
    # +allow_nil:+ - If true, allows the mapped value to be nil. Defaults to true.
    #
    # == Example
    #
    #    class Foo < LazyMapper::Model
    #      one :boss, Person, from: "supervisor", map: ->(p) { Person.new(p) }
    #      one :weapon, [BladedWeapon, Firearm], default: Sixshooter.new
    #      # ...
    #    end
    #
    def self.one(name, type, from: map_name(name), allow_nil: true, **args)

      ivar = IVAR[name]

      # Define writer
      define_method(WRITER[name]) { |val|
        check_type! val, type, allow_nil: allow_nil
        instance_variable_set(ivar, val)
      }

      # Define reader
      define_method(name) {
        memoize(name, ivar) {
          unmapped_value = unmapped_data[from]
          mapped_value(name, unmapped_value, type, **args)
        }
      }

      attributes[name] = type
    end

    #
    # Converts a value to +true+ or +false+ according to its truthyness
    #
    TO_BOOL = -> b { !!b }

    #
    # Defines an boolean attribute
    #
    # == Arguments
    #
    # +name+ - The name of the attribue
    #
    # +from:+ - Specifies the name of the wrapped value in the JSON object.
    # Defaults to camelCased version of +name+.
    #
    # +map:+ - Specifies a custom mapper to apply to the wrapped value. Must be a Callable.
    # Defaults to TO_BOOL if unspecified.
    #
    # +default:+ The default value to use if the value is missing. False, if unspecified
    #
    # == Example
    #
    #    class Foo < LazyMapper::Model
    #      is :green?, from: "isGreen", map: ->(x) { !x.zero? }
    #      # ...
    #    end
    #
    def self.is name, from: map_name(name), map: TO_BOOL, default: false
      one name, [TrueClass, FalseClass], from: from, allow_nil: false, map: map, default: default
    end

    singleton_class.send(:alias_method, :has, :is)

    #
    # Defines a collection attribute
    #
    # == Arguments
    #
    # +name+ - The name of the attribute
    #
    # +type+ - The type of the elements in the collection.
    #
    # +from:+ - Specifies the name of the wrapped array in the unmapped data.
    # Defaults to camelCased version of +name+.
    #
    # +map:+ - Specifies a custom mapper to apply to each elements in the wrapped collection.
    # If unspecified, it defaults to the default mapper for the specified +type+ or simply the identity mapper
    # if no default mapper exists.
    #
    # +default:+ - The default value to use, if the unmapped value is missing.
    #
    # == Example
    #
    #    class Bar < LazyMapper::Model
    #      many :underlings, Person, from: "serfs", map: ->(p) { Person.new(p) }
    #      # ...
    #    end
    #
    def self.many(name, type, from: map_name(name), **args)

      # Define setter
      define_method(WRITER[name]) { |val|
        check_type! val, Enumerable, allow_nil: false
        instance_variable_set(IVAR[name], val)
      }

      # Define getter
      define_method(name) {
        memoize(name) {
          unmapped_value = unmapped_data[from]
          if unmapped_value.is_a? Array
            unmapped_value.map { |v| mapped_value(name, v, type, **args) }
          else
            mapped_value name, unmapped_value, Array, **args
          end
        }
      }

      attributes[name] = Array
    end

    #
    # Adds an instance-level type mapper
    #
    def add_mapper_for(type, &block)
      mappers[type] = block
    end

    #
    # Returns a +Hash+ with keys corresponding to attribute names, and values
    # corresponding to *mapped* attribute values.
    #
    # Note: This will eagerly map all attributes that haven't yet been mapped
    #
    def to_h
      attributes.each_with_object({}) { |(key, _value), h|
        h[key] = self.send key
      }
    end

    def inspect
      @__under_inspection__ ||= 0
      return "<#{ self.class.name } ... >" if @__under_inspection__.positive?

      @__under_inspection__ += 1
      present_attributes = attributes.keys.each_with_object({}) { |name, memo|
        ivar = IVAR[name]
        next unless self.instance_variable_defined? ivar

        memo[name] = self.instance_variable_get ivar
      }

      res = "<#{ self.class.name } #{ present_attributes.map { |k, v| k.to_s + ': ' + v.inspect }.join(', ') } >"
      @__under_inspection__ -= 1
      res
    end

    #
    # Defines how to map an attribute name
    # to the corresponding name in the unmapped
    # JSON object.
    #
    # Defaults to CAMELIZE
    #
    def self.map_name(name)
      CAMELIZE[name]
    end

    private

    attr_writer :unmapped_data
    attr_writer :mappers

    def unmapped_data
      @unmapped_data ||= {}
    end

    def mapping_for(name, type)
      mappers[name] || mappers[type] || self.class.mappers[type]
    end

    def default_value(type)
      self.class.default_values[type]
    end

    def attributes
      self.class.attributes
    end

    def mapped_value(name, unmapped_value, type, map: mapping_for(name, type), default: default_value(type))
      return default.dup if unmapped_value.nil? # Duplicate to prevent accidental sharing between instances

      if map.nil?
        fail ArgumentError, "missing mapper for #{ name } (#{ type }). "\
          "Unmapped value: #{ unmapped_value.inspect }"
      end

      return map.call(unmapped_value, self) if map.arity > 1

      map.call(unmapped_value)
    end

    def check_type! value, type, allow_nil:
      permitted_types = allow_nil ? Array(type) + [ NilClass ] : Array(type)
      return if permitted_types.any? value.method(:is_a?)

      fail TypeError.new "#{ self.class.name }: "\
        "#{ value.inspect } is a #{ value.class } "\
        "but was supposed to be a #{ humanize_list permitted_types, conjunction: ' or ' }"
    end

    # [1,2,3] -> "1, 2 and 3"
    # [1, 2]  -> "1 and 2"
    # [1]     -> "1"
    def humanize_list terms, separator: ', ', conjunction: ' and '
      *all_but_last, last = terms
      return last if all_but_last.empty?

      [ all_but_last.join(separator), last ].join conjunction
    end

    def memoize name, ivar = IVAR[name]
      send WRITER[name], yield unless instance_variable_defined?(ivar)
      instance_variable_get(ivar)
    end

    SNAKE_CASE_PATTERN = /(_[a-z])/ # :nodoc:
    CAMELIZE = -> name { name.to_s.gsub(SNAKE_CASE_PATTERN) { |x| x[1].upcase }.delete('?') }

  end
end
